### PR TITLE
issue-278

### PR DESCRIPTION
**Pull request by syedowaisali90**
Hi @vincent-karuri I have created the for review.

### CODE REVIEW

File: rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/activity/CovidPatientRegisterActivity.java [hunk 1]
```diff
@@ -18,6 +27,14 @@
  */
 public class CovidPatientRegisterActivity extends PatientRegisterActivity {
 
+    private static final String LOCALE_IN = "in";
```
Comment (vincent-karuri) [inline rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/activity/CovidPatientRegisterActivity.java:30]: I think this should be in `CovidConstants`?

File: rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/activity/CovidPatientRegisterActivity.java [hunk 2]
```diff
@@ -49,4 +70,55 @@ protected Class getLoginPage() {
     protected PatientRegisterActivityPresenter createPatientRegisterActivityPresenter() {
         return new CovidPatientRegisterActivityPresenter(this);
     }
+
+    private Map<String, Locale> locales = new HashMap<>();
+    private String[] languages;
+    private int currentLanguageIndex = 0;
+
```
Comment (vincent-karuri) [inline rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/activity/CovidPatientRegisterActivity.java:77]: Global variables should be at the top of the file.

File: rdt-open-reader/src/main/java/io/ona/rdt/util/Utils.java [hunk 3]
```diff
@@ -158,4 +162,13 @@ public static boolean tableExists(SQLiteDatabase db, String tableName) {
     public static boolean isEmptyCursor(Cursor cursor) {
         return cursor == null || cursor.getCount() == 0;
     }
+
+    public static AllSharedPreferences getSharedPreference(Context context) {
+        return new AllSharedPreferences(PreferenceManager.getDefaultSharedPreferences(context)) {
+            @Override
+            public String fetchLanguagePreference() {
+                return getPreferences().getString("locale", BuildConfig.LOCALE).trim();
```
Comment (vincent-karuri) [inline rdt-open-reader/src/main/java/io/ona/rdt/util/Utils.java:170]: `RDTApplication.getInstance().getContext().allSharedPreferences().getPreference(KEY);` would work for getting a saved preference.

File: rdt-open-reader/src/main/java/io/ona/rdt/util/Utils.java [hunk 4]
```diff
@@ -99,8 +102,9 @@ public static int convertDpToPixels(Context context, float dp) {
         return (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dp, context.getResources().getDisplayMetrics());
     }
 
-    public static void updateLocale(android.content.Context context) {
-        Locale locale = new Locale(BuildConfig.LOCALE);
+    public static void updateLocale(android.content.Context context, AllSharedPreferences allSharedPreferences) {
+
+        Locale locale = new Locale(allSharedPreferences.fetchLanguagePreference());
```
Comment (vincent-karuri) [inline rdt-open-reader/src/main/java/io/ona/rdt/util/Utils.java:107]: https://github.com/onaio/rdt-standard/pull/457/files#r478863684. No need to pass the `AllSharedPreferences`.

File: rdt-open-reader/src/main/java/io/ona/rdt/activity/RDTJsonFormActivity.java [hunk 5]
```diff
@@ -31,7 +33,7 @@
 
     @Override
     protected void onCreate(Bundle savedInstanceState) {
-        updateLocale(this);
+        updateLocale(this, Utils.getSharedPreference(this));
```
Comment (vincent-karuri) [inline rdt-open-reader/src/main/java/io/ona/rdt/activity/RDTJsonFormActivity.java:36]: No need to pass `Utils.getSharedPreference(this)` here based on https://github.com/onaio/rdt-standard/pull/457/files#r478863684.

File: rdt-open-reader/src/main/java/io/ona/rdt/activity/RDTExpirationDateActivity.java [hunk 6]
```diff
@@ -19,7 +20,7 @@
 
     @Override
     public void onCreate(Bundle savedInstanceState) {
-        updateLocale(this);
+        updateLocale(this, Utils.getSharedPreference(this));
```
Comment (vincent-karuri) [inline rdt-open-reader/src/main/java/io/ona/rdt/activity/RDTExpirationDateActivity.java:23]: https://github.com/onaio/rdt-standard/pull/457/files#r478864378

File: rdt-open-reader/src/main/java/io/ona/rdt/activity/PatientRegisterActivity.java [hunk 7]
```diff
@@ -48,7 +49,7 @@
 
     @Override
     protected void onCreate(Bundle savedInstanceState) {
-        updateLocale(this);
+        updateLocale(this, Utils.getSharedPreference(this));
```
Comment (vincent-karuri) [inline rdt-open-reader/src/main/java/io/ona/rdt/activity/PatientRegisterActivity.java:52]: https://github.com/onaio/rdt-standard/pull/457/files#r478864378

File: rdt-open-reader/src/main/java/io/ona/rdt/activity/PatientProfileActivity.java [hunk 8]
```diff
@@ -6,26 +6,28 @@
...
 
-public class PatientProfileActivity extends FragmentActivity implements PatientProfileActivityContract.View {
+public class PatientProfileActivity extends MultiLanguageActivity implements PatientProfileActivityContract.View {
 
     protected PatientProfileActivityPresenter presenter;
 
     @Override
     protected void onCreate(Bundle savedInstanceState) {
-        updateLocale(this);
+        updateLocale(this, Utils.getSharedPreference(this));
```
Comment (vincent-karuri) [inline rdt-open-reader/src/main/java/io/ona/rdt/activity/PatientProfileActivity.java:30]: https://github.com/onaio/rdt-standard/pull/457/files#r478864378

File: rdt-open-reader/src/main/java/io/ona/rdt/activity/PatientProfileActivity.java [hunk 9]
```diff
@@ -6,26 +6,28 @@
...
 import io.ona.rdt.presenter.PatientProfileActivityPresenter;
+import io.ona.rdt.util.Utils;
 import timber.log.Timber;
 
 import static io.ona.rdt.util.Constants.FormFields.PATIENT;
 import static io.ona.rdt.util.Constants.RequestCodes.REQUEST_CODE_GET_JSON;
 import static io.ona.rdt.util.Utils.updateLocale;
 
-public class PatientProfileActivity extends FragmentActivity implements PatientProfileActivityContract.View {
+public class PatientProfileActivity extends MultiLanguageActivity implements PatientProfileActivityContract.View {
```
Comment (vincent-karuri) [inline rdt-open-reader/src/main/java/io/ona/rdt/activity/PatientProfileActivity.java:24]: Any reason to extend `MultiLanguageActivity` here?

File: rdt-open-reader/src/main/java/io/ona/rdt/activity/LoginActivity.java [hunk 10]
```diff
@@ -24,7 +25,7 @@
 
     @Override
     public void onCreate(Bundle savedInstanceState) {
-        updateLocale(this);
+        updateLocale(this, Utils.getSharedPreference(this));
```
Comment (vincent-karuri) [inline rdt-open-reader/src/main/java/io/ona/rdt/activity/LoginActivity.java:28]: https://github.com/onaio/rdt-standard/pull/457/files#r478864378.

File: rdt-open-reader/src/main/java/io/ona/rdt/activity/CustomRDTCaptureActivity.java [hunk 11]
```diff
@@ -37,7 +38,7 @@
 
     @Override
     protected void onCreate(Bundle savedInstanceState) {
-        updateLocale(this);
+        updateLocale(this, Utils.getSharedPreference(this));
```
Comment (vincent-karuri) [inline rdt-open-reader/src/main/java/io/ona/rdt/activity/CustomRDTCaptureActivity.java:41]: https://github.com/onaio/rdt-standard/pull/457/files#r478864378

File: rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/activity/CovidPatientRegisterActivity.java [hunk 12]
```diff
@@ -49,4 +70,55 @@ protected Class getLoginPage() {
     protected PatientRegisterActivityPresenter createPatientRegisterActivityPresenter() {
         return new CovidPatientRegisterActivityPresenter(this);
     }
+
+    private Map<String, Locale> locales = new HashMap<>();
+    private String[] languages;
```
Comment (vincent-karuri) [inline rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/activity/CovidPatientRegisterActivity.java:75]: Not clear to me why we need to keep a map and an array, can we just have the array if that's what the `AlertDialog` requires?

Comment (vincent-karuri) [reply rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/activity/CovidPatientRegisterActivity.java:75]: We can always build the Locale from the language code saved in `SharedPreferences` if need be.

File: rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/activity/CovidPatientRegisterActivity.java (hunk 1 above)
Comment (vincent-karuri) [reply rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/activity/CovidPatientRegisterActivity.java:30]: Also, although `in` might work, the new language code for Bahasa is `id`.

**Review by vincent-karuri — CHANGES_REQUESTED** [review]
Please address the comments.

File: rdt-open-reader/src/main/java/io/ona/rdt/util/Utils.java (hunk 4 above)
Comment (vincent-karuri) [reply rdt-open-reader/src/main/java/io/ona/rdt/util/Utils.java:107]: We may also need to have a default fallback Locale that can be defined in the gradle file.

File: rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/activity/CovidPatientRegisterActivity.java (hunk 1 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/activity/CovidPatientRegisterActivity.java:30]: done.

File: rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/activity/CovidPatientRegisterActivity.java (hunk 2 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/activity/CovidPatientRegisterActivity.java:77]: done.

File: rdt-open-reader/src/main/java/io/ona/rdt/util/Utils.java (hunk 3 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/main/java/io/ona/rdt/util/Utils.java:170]: done.

File: rdt-open-reader/src/main/java/io/ona/rdt/util/Utils.java (hunk 4 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/main/java/io/ona/rdt/util/Utils.java:107]: done

File: rdt-open-reader/src/main/java/io/ona/rdt/activity/RDTExpirationDateActivity.java (hunk 6 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/main/java/io/ona/rdt/activity/RDTExpirationDateActivity.java:23]: done.

File: rdt-open-reader/src/main/java/io/ona/rdt/activity/PatientRegisterActivity.java (hunk 7 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/main/java/io/ona/rdt/activity/PatientRegisterActivity.java:52]: done.

File: rdt-open-reader/src/main/java/io/ona/rdt/activity/PatientProfileActivity.java (hunk 8 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/main/java/io/ona/rdt/activity/PatientProfileActivity.java:30]: done.

File: rdt-open-reader/src/main/java/io/ona/rdt/activity/PatientProfileActivity.java (hunk 9 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/main/java/io/ona/rdt/activity/PatientProfileActivity.java:24]: just for testing purpose now back to extends `FragmentActivity`

File: rdt-open-reader/src/main/java/io/ona/rdt/activity/LoginActivity.java (hunk 10 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/main/java/io/ona/rdt/activity/LoginActivity.java:28]: done.

File: rdt-open-reader/src/main/java/io/ona/rdt/activity/CustomRDTCaptureActivity.java (hunk 11 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/main/java/io/ona/rdt/activity/CustomRDTCaptureActivity.java:41]: done.

File: rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/activity/CovidPatientRegisterActivity.java (hunk 12 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/activity/CovidPatientRegisterActivity.java:75]: done. now i have used `string-array`

File: rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/util/CovidConstants.java [hunk 13]
```diff
@@ -88,4 +88,8 @@
     interface RequestCodes {
         int LOCATION_PERMISSIONS = 8732;
     }
+
+    interface Language {
+        String INDONESIA = "id";
```
Comment (vincent-karuri) [inline rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/util/CovidConstants.java:93]: Is this constant in use?

File: rdt-open-reader/src/main/java/io/ona/rdt/util/Utils.java [hunk 14]
```diff
@@ -102,9 +100,11 @@ public static int convertDpToPixels(Context context, float dp) {
         return (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dp, context.getResources().getDisplayMetrics());
     }
 
-    public static void updateLocale(android.content.Context context, AllSharedPreferences allSharedPreferences) {
+    public static void updateLocale(Context context) {
 
-        Locale locale = new Locale(allSharedPreferences.fetchLanguagePreference());
+        String lang = RDTApplication.getInstance().getContext().allSharedPreferences().getPreference("locale");
```
Comment (vincent-karuri) [inline rdt-open-reader/src/main/java/io/ona/rdt/util/Utils.java:105]: "locale" can move to CovidConstants.

File: rdt-open-reader/src/main/java/io/ona/rdt/util/Utils.java [hunk 15]
```diff
@@ -102,9 +100,11 @@ public static int convertDpToPixels(Context context, float dp) {
...
+
+        Locale locale = new Locale(lang.length() > 0 ? lang : BuildConfig.LOCALE);
```
Comment (vincent-karuri) [inline rdt-open-reader/src/main/java/io/ona/rdt/util/Utils.java:107]: `lang` can potentially be null here, use `StringUtils.isBlank(lang)` instead.

File: rdt-open-reader/src/main/java/io/ona/rdt/activity/LoginActivity.java [hunk 16]
```diff
@@ -43,7 +43,6 @@ protected String getRegisterTableName() {
     }
 
     private void addAttributionText() {
-        updateLocale(this);
```
Comment (vincent-karuri) [inline rdt-open-reader/src/main/java/io/ona/rdt/activity/LoginActivity.java:46]: Any reason this is removed?

File: rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/util/CovidConstants.java (hunk 13 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/covidIndonesia/java/io/ona/rdt/util/CovidConstants.java:93]: not, removed.

File: rdt-open-reader/src/main/java/io/ona/rdt/util/Utils.java (hunk 14 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/main/java/io/ona/rdt/util/Utils.java:105]: moved to CovidConstants.

File: rdt-open-reader/src/main/java/io/ona/rdt/util/Utils.java (hunk 15 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/main/java/io/ona/rdt/util/Utils.java:107]: fixed.

File: rdt-open-reader/src/main/java/io/ona/rdt/activity/LoginActivity.java (hunk 16 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/main/java/io/ona/rdt/activity/LoginActivity.java:46]: added, remove by mistake

File: rdt-open-reader/src/main/java/io/ona/rdt/util/Utils.java (hunk 15 above)
Comment (vincent-karuri) [reply rdt-open-reader/src/main/java/io/ona/rdt/util/Utils.java]: Sorry, the comment here was to change this line to `Locale locale = new Locale(StringUtils.isNotBlank(lang) ? lang : BuildConfig.LOCALE);`.

**Comment by vincent-karuri** [comment]
Please add a test to verify the language switcher sets the correct locale.

File: rdt-open-reader/src/main/java/io/ona/rdt/util/Utils.java (hunk 15 above)
Comment (syedowaisali90) [reply rdt-open-reader/src/main/java/io/ona/rdt/util/Utils.java]: fixed.

**Pull request merged**